src/memory.h

```diff
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: GPL-2.0 OR BSD-2-Clause */
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#ifndef MEMORY_H
+#define MEMORY_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#define container_of(ptr, type, member) ({                      \
+                const typeof( ((type *)0)->member ) *__mptr = (ptr);    \
+                (type *)( (char *)__mptr - offsetof(type,member) );})
+
+
+/* Base context for memory management to be extended by concrete implementations */
+struct memory_ctx {
+	int (*init)(struct memory_ctx *ctx);
+	int (*destroy)(struct memory_ctx *ctx);
+	int (*allocate_buffer)(struct memory_ctx *ctx, int alignment, uint64_t size, int *dmabuf_fd, void **addr, bool *can_init);
```

Comment:
You might also need a copy hook, to copy data to and from the device. I don't think the current benchmarks use it, but if you wanted to add GPU support to, say, ib_write_lat... we would need a copy from the device to the host on the target side to check completions.

Reply:
Ack,  on this and the other comments about avoiding functional changes in a commit meant to refactor code.